crates/sui-core/src/authority_active/gossip/node_sync.rs

```diff
@@ -0,0 +1,435 @@
+
+            self.effects_stake_map
+                .entry(*digests)
+                .and_modify(|cur| *cur += stake)
+                .or_insert(stake)
+        };
+
+        let is_final = *cur_stake >= quorum_threshold;
+        if !is_final {
+            trace!(
+                ?digests,
+                "tx cert/effects not yet final: {} < {}",
+                *cur_stake,
+                quorum_threshold
+            );
+        }
+        is_final
+    }
+}
+
+/// Waiter is used to single-shot concurrent requests and wait for dependencies to finish.
+struct Waiter<Key, ResultT> {
+    waiters: Mutex<HashMap<Key, broadcast::Sender<ResultT>>>,
+}
+
+impl<Key: std::hash::Hash + Eq + Clone + std::fmt::Debug, ResultT: Clone> Waiter<Key, ResultT> {
+    fn new() -> Self {
+        Self {
+            waiters: Mutex::new(HashMap::new()),
+        }
+    }
+
+    /// Returns (Some(tx), rx) if there are no other waiters yet, or else (None, rx).
+    /// All rxes can be woken by sending to the supplied tx, or by calling notify(key, result)
+    async fn wait(
+        &self,
+        key: &Key,
+    ) -> (
+        Option<broadcast::Sender<ResultT>>,
+        broadcast::Receiver<ResultT>,
+    ) {
+        let waiters = &mut self.waiters.lock().await;
+        let entry = waiters.entry(key.clone());
+
+        match entry {
+            hash_map::Entry::Occupied(e) => (None, e.get().subscribe()),
+            hash_map::Entry::Vacant(e) => {
+                let (tx, rx) = broadcast::channel(1);
+                e.insert(tx.clone());
+                (Some(tx), rx)
+            }
+        }
+    }
+
+    async fn notify(&self, key: &Key, res: ResultT) -> SuiResult {
+        if let Some(tx) = self.waiters.lock().await.remove(key) {
+            tx.send(res).map_err(|_| SuiError::GenericAuthorityError {
+                error: format!("couldn't notify waiters for key {:?}", key),
+            })?;
+        }
+        // else: no one was waiting on this key.
+        Ok(())
+    }
+}
+
+struct DigestsMessage {
+    digests: ExecutionDigests,
+    peer: AuthorityName,
+}
+
+/// NodeSyncState is shared by any number of NodeSyncDigestHandler's, and receives DigestsMessage
+/// messages from those handlers, waits for finality of TXes, and then downloads and applies those
+/// TXes locally.
+struct NodeSyncState<A> {
+    committee: Arc<Committee>,
+    effects_stake: Mutex<EffectsStakeMap>,
+    state: Arc<AuthorityState>,
+    node_sync_store: Arc<NodeSyncStore>,
+    aggregator: Arc<AuthorityAggregator<A>>,
+
+    // Used to single-shot multiple concurrent downloads.
+    pending_downloads: Waiter<TransactionDigest, SuiResult>,
+
+    // Used to wait for parent transactions to be applied locally
+    pending_txes: Waiter<TransactionDigest, ()>,
+}
+
+impl<A> NodeSyncState<A>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    fn start(self_: Arc<Self>, mut receiver: mpsc::Receiver<DigestsMessage>) -> JoinHandle<()> {
```

Comment:
```suggestion
    fn start(self: Arc<Self>, mut receiver: mpsc::Receiver<DigestsMessage>) -> JoinHandle<()> {
```

drop the underscore when naming "self"

Reply:
Although I don't understand why you need to have things in here wrapped in Arcs and Mutexes if you're spawning this thing inside its own task. I'd instead expect you to use message passing...which you also seem to do.

Reply:
Read a little further down... We spawn a task for each digest we receive so that we can download certs concurrently. Every task needs a an Arc<NodeSyncState>.

(Without this, a node can never sync faster than 1 cert per RTT, which could easily be as low as 5-10 per second, so I don't think this is a premature optimization at all).

Reply:
Also thanks re `self_`, I thought that `self: Arc<Self>` was an unstable feature.